File: backend/app/ml/risk_engine.py
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.preprocessing import StandardScaler
from typing import Dict, Any, List, Tuple
from ..schemas.risk import (
    RiskEvaluationInput,
    RiskScoreResultSchema,
    RiskFactorSchema,
    CustomerBaselineSchema,
    RiskMetricsSchema,
    SimulationPayloadSchema,
)

class ScikitRiskEngine:
    def __init__(self):
        self.model = RandomForestClassifier(n_estimators=100, max_depth=6, random_state=42)
        self.scaler = StandardScaler()
        self.is_trained = False
        self.feature_names = [
            "amount_ratio",
            "velocity_10m",
            "is_new_device",
            "is_location_mismatch",
            "failed_attempts",
            "is_new_merchant",
            "account_age_days",
            "is_odd_hour"
        ]
        self._train_initial_model()

    def _train_initial_model(self):
        """Train the Scikit-learn model on synthetic training distributions."""
        np.random.seed(42)
        n_samples = 2500

        # Normal benign distribution (80%)
        n_normal = int(n_samples * 0.82)
        normal_amount_ratio = np.random.uniform(0.3, 2.2, n_normal)
        normal_velocity = np.random.choice([1, 2, 3], size=n_normal, p=[0.75, 0.20, 0.05])
        normal_new_device = np.random.choice([0, 1], size=n_normal, p=[0.92, 0.08])
        normal_location_mismatch = np.random.choice([0, 1], size=n_normal, p=[0.95, 0.05])
        normal_failed_attempts = np.random.choice([0, 1], size=n_normal, p=[0.96, 0.04])
        normal_new_merchant = np.random.choice([0, 1], size=n_normal, p=[0.85, 0.15])
        normal_account_age = np.random.uniform(60, 1200, n_normal)
        normal_odd_hour = np.random.choice([0, 1], size=n_normal, p=[0.90, 0.10])
        normal_labels = np.zeros(n_normal)

        # Anomalous fraud distribution (18%)
        n_fraud = n_samples - n_normal
        fraud_amount_ratio = np.random.uniform(4.0, 35.0, n_fraud)
        fraud_velocity = np.random.randint(4, 18, n_fraud)
        fraud_new_device = np.random.choice([0, 1], size=n_fraud, p=[0.15, 0.85])
        fraud_location_mismatch = np.random.choice([0, 1], size=n_fraud, p=[0.20, 0.80])
        fraud_failed_attempts = np.random.randint(2, 7, n_fraud)
        fraud_new_merchant = np.random.choice([0, 1], size=n_fraud, p=[0.30, 0.70])
        fraud_account_age = np.random.uniform(1, 90, n_fraud)
        fraud_odd_hour = np.random.choice([0, 1], size=n_fraud, p=[0.40, 0.60])
        fraud_labels = np.ones(n_fraud)

        X = np.column_stack([
            np.concatenate([normal_amount_ratio, fraud_amount_ratio]),
            np.concatenate([normal_velocity, fraud_velocity]),
            np.concatenate([normal_new_device, fraud_new_device]),
            np.concatenate([normal_location_mismatch, fraud_location_mismatch]),
            np.concatenate([normal_failed_attempts, fraud_failed_attempts]),
            np.concatenate([normal_new_merchant, fraud_new_merchant]),
            np.concatenate([normal_account_age, fraud_account_age]),
            np.concatenate([normal_odd_hour, fraud_odd_hour]),
        ])
        y = np.concatenate([normal_labels, fraud_labels])

        X_scaled = self.scaler.fit_transform(X)
        self.model.fit(X_scaled, y)
        self.is_trained = True

    def extract_features(self, inp: RiskEvaluationInput) -> Tuple[np.ndarray, Dict[str, Any]]:
        avg_amt = max(inp.averageCustomerTransaction or 2500.0, 100.0)
        amount_ratio = inp.amount / avg_amt
        is_odd_hour = 1 if (inp.timeOfDayHour is not None and 1 <= inp.timeOfDayHour <= 4) else 0
        is_loc_mismatch = 1 if (inp.previousLocation and inp.location != inp.previousLocation) else 0
        
        feature_vector = np.array([[
            amount_ratio,
            inp.velocityLast10m,
            1 if inp.isNewDevice else 0,
            is_loc_mismatch,
            inp.failedAttemptsLast24h,
            1 if inp.isNewMerchant else 0,
            inp.accountAgeDays or 365,
            is_odd_hour
        ]])

        context = {
            "amount_ratio": amount_ratio,
            "avg_amount": avg_amt,
            "is_location_mismatch": bool(is_loc_mismatch),
            "is_odd_hour": bool(is_odd_hour),
        }
        return feature_vector, context

    def evaluate(
        self,
        inp: RiskEvaluationInput,
        weights: Dict[str, int] = None,
        thresholds: Dict[str, int] = None
    ) -> RiskScoreResultSchema:
        if weights is None:
            weights = {
                "amountDeviationWeight": 22,
                "newDeviceWeight": 18,
                "locationMismatchWeight": 15,
                "velocityWeight": 20,
                "failedAttemptsWeight": 16,
                "newMerchantWeight": 10,
                "highRiskCategoryWeight": 12,
                "timeAnomalyWeight": 8,
            }
        if thresholds is None:
            thresholds = {
                "approveMax": 30,
                "verifyMax": 60,
                "holdMax": 80,
                "blockMin": 81,
            }

        feat_vec, ctx = self.extract_features(inp)
        feat_scaled = self.scaler.transform(feat_vec)
        ml_prob = float(self.model.predict_proba(feat_scaled)[0][1])

        # Decompose factor contributions
        factors: List[RiskFactorSchema] = []
        raw_score = 5 + int(ml_prob * 15)  # Baseline network risk

        # 1. Amount ratio
        amt_ratio = ctx["amount_ratio"]
        if amt_ratio >= 20.0:
            contrib = min(30, int(weights["amountDeviationWeight"] * 1.3))
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="AMOUNT_MASSIVE_SPIKE",
                name=f"Amount {amt_ratio:.0f}× customer average",
                contribution=contrib,
                description=f"Payment (₹{inp.amount:,.0f}) is {amt_ratio:.0f}x higher than historical average ₹{ctx['avg_amount']:,.0f}.",
                category="AMOUNT",
                severity="CRITICAL"
            ))
        elif amt_ratio >= 5.0:
            contrib = min(22, int(weights["amountDeviationWeight"] * 0.9))
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="AMOUNT_HIGH_SPIKE",
                name=f"Amount {amt_ratio:.1f}× customer average",
                contribution=contrib,
                description=f"Payment (₹{inp.amount:,.0f}) significantly exceeds historical spending average.",
                category="AMOUNT",
                severity="HIGH"
            ))
        elif amt_ratio >= 2.5:
            contrib = int(weights["amountDeviationWeight"] * 0.45)
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="AMOUNT_MODERATE_SPIKE",
                name="Higher than usual amount",
                contribution=contrib,
                description="Payment amount is moderately higher than usual baseline.",
                category="AMOUNT",
                severity="MEDIUM"
            ))

        # 2. Device novelty
        if inp.isNewDevice:
            contrib = weights["newDeviceWeight"]
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="DEVICE_UNRECOGNIZED",
                name="New / Unrecognized device fingerprint",
                contribution=contrib,
                description="Transaction originated from a previously unseen hardware/browser signature.",
                category="DEVICE",
                severity="HIGH"
            ))

        # 3. Location anomaly
        if ctx["is_location_mismatch"]:
            contrib = weights["locationMismatchWeight"]
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="LOCATION_ANOMALY",
                name="Unusual / Mismatched location",
                contribution=contrib,
                description=f"Originating city ({inp.location}) deviates from customer established activity hub.",
                category="LOCATION",
                severity="HIGH"
            ))

        # 4. Velocity burst
        if inp.velocityLast10m >= 10:
            contrib = min(30, int(weights["velocityWeight"] * 1.4))
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="VELOCITY_EXTREME",
                name="Critical transaction velocity burst",
                contribution=contrib,
                description=f"{inp.velocityLast10m} payment attempts recorded in the last 10 minutes (Bot/carding signature).",
                category="VELOCITY",
                severity="CRITICAL"
            ))
        elif inp.velocityLast10m >= 4:
            contrib = weights["velocityWeight"]
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="VELOCITY_HIGH",
                name="High transaction velocity",
                contribution=contrib,
                description=f"{inp.velocityLast10m} transactions initiated in a short 10-minute window.",
                category="VELOCITY",
                severity="HIGH"
            ))

        # 5. Failed attempts
        if inp.failedAttemptsLast24h >= 4:
            contrib = min(25, int(weights["failedAttemptsWeight"] * 1.3))
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="FAILED_ATTEMPTS_SPIKE",
                name=f"{inp.failedAttemptsLast24h} recent authorization failures",
                contribution=contrib,
                description="Repeated CVV/OTP declines noted prior to this checkout attempt.",
                category="BEHAVIOR",
                severity="HIGH"
            ))
        elif inp.failedAttemptsLast24h >= 2:
            contrib = int(weights["failedAttemptsWeight"] * 0.6)
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="FAILED_ATTEMPTS_MODERATE",
                name=f"{inp.failedAttemptsLast24h} recent declines",
                contribution=contrib,
                description="Prior authorization failures recorded in the last 24 hours.",
                category="BEHAVIOR",
                severity="MEDIUM"
            ))

        # 6. First time merchant
        if inp.isNewMerchant:
            contrib = weights["newMerchantWeight"]
            raw_score += contrib
            factors.append(RiskFactorSchema(
                code="MERCHANT_FIRST_TIME",
                name="First-time merchant interaction",
                contribution=contrib,
                description="Customer has no prior payment history with this merchant.",
                category="MERCHANT",
                severity="LOW"
            ))

        # Final score bounded 0-100
        final_score = min(100.0, max(1.0, float(raw_score)))

        # Decision thresholds
        if final_score <= thresholds["approveMax"]:
            risk_level = "LOW"
            decision = "APPROVE"
        elif final_score <= thresholds["verifyMax"]:
            risk_level = "MEDIUM"
            decision = "VERIFY"
        elif final_score <= thresholds["holdMax"]:
            risk_level = "HIGH"
            decision = "HOLD"
        else:
            risk_level = "CRITICAL"
            decision = "BLOCK"

        # Generate Explainable AI narrative
        explanation = self._generate_explanation(
            final_score, decision, amt_ratio, inp.amount, ctx["avg_amount"],
            inp.isNewDevice, ctx["is_location_mismatch"], inp.location,
            inp.velocityLast10m, inp.failedAttemptsLast24h, inp.isNewMerchant
        )

        return RiskScoreResultSchema(
            riskScore=final_score,
            riskLevel=risk_level,
            decision=decision,
            riskFactors=sorted(factors, key=lambda x: x.contribution, reverse=True),
            explanation=explanation,
            customerBaseline=CustomerBaselineSchema(
                averageAmount=ctx["avg_amount"],
                typicalRange=[round(ctx["avg_amount"] * 0.3), round(ctx["avg_amount"] * 2.2)],
                usualLocation=inp.previousLocation or "Hyderabad",
                usualDevice="Android Device",
                typicalDailyTxn=3
            ),
            metrics=RiskMetricsSchema(
                amountRatio=amt_ratio,
                velocityLast10m=inp.velocityLast10m,
                isNewDevice=inp.isNewDevice,
                isLocationAnomaly=ctx["is_location_mismatch"],
                failedAttempts=inp.failedAttemptsLast24h,
                isNewMerchant=inp.isNewMerchant,
                mlPredictionProbability=round(ml_prob, 3)
            ),
            modelType="Scikit-Learn Random Forest Classifier + Feature Attribution Engine"
        )

    def _generate_explanation(
        self, score: float, decision: str, amt_ratio: float, amount: float,
        avg_amt: float, is_new_device: bool, is_loc_mismatch: bool,
        location: str, velocity: int, failed: int, is_new_merch: bool
    ) -> str:
        if score <= 30:
            return (
                f"This transaction aligns closely with the customer's typical behavioral baseline. "
                f"The transaction amount (₹{amount:,.0f}) is within expected limits, initiated from a recognized device "
                f"in {location} with normal velocity. Risk level is low and safe to auto-approve."
            )

        reasons = []
        if amt_ratio >= 5.0:
            reasons.append(f"the payment amount (₹{amount:,.0f}) is significantly higher ({amt_ratio:.1f}×) than customer historical average of ₹{avg_amt:,.0f}")
        elif amt_ratio >= 2.0:
            reasons.append("the payment amount is moderately elevated compared to past transactions")

        if is_new_device:
            reasons.append("the transaction originated from a new, unrecognized device fingerprint")
        if is_loc_mismatch:
            reasons.append(f"the originating location ({location}) differs from customer typical activity center")
        if velocity >= 4:
            reasons.append(f"{velocity} rapid transactions occurred within the last 10 minutes")
        if failed >= 2:
            reasons.append(f"there were {failed} prior failed authorization attempts in the last 24 hours")
        if is_new_merch:
            reasons.append("this is the first transaction with this merchant")

        if not reasons:
            return f"This transaction presents statistical anomaly features warranting a risk score of {score:.0f}/100 and a recommendation to {decision}."

        if len(reasons) == 1:
            joined = reasons[0]
        else:
            joined = ", ".join(reasons[:-1]) + ", and " + reasons[-1]

        return f"This transaction was flagged with a risk score of {score:.0f}/100 because {joined}. The system recommends {decision} pending risk analyst verification."

risk_engine = ScikitRiskEngine()

File: src/engine/riskEngine.ts
import {
  DecisionType,
  RiskLevel,
  RiskFactor,
  RiskScoreResult,
  EngineThresholds,
  EngineWeights,
  SimulationPayload,
  Customer,
  Merchant
} from '../types';

export const DEFAULT_THRESHOLDS: EngineThresholds = {
  approveMax: 30,
  verifyMax: 60,
  holdMax: 80,
  blockMin: 81,
};

export const DEFAULT_WEIGHTS: EngineWeights = {
  amountDeviationWeight: 22,
  newDeviceWeight: 18,
  locationMismatchWeight: 15,
  velocityWeight: 20,
  failedAttemptsWeight: 16,
  newMerchantWeight: 10,
  highRiskCategoryWeight: 12,
  timeAnomalyWeight: 8,
};

export interface RiskEvaluationInput {
  amount: number;
  customerId: string;
  merchantId: string;
  merchantCategory?: string;
  location: string;
  previousLocation?: string;
  isNewDevice: boolean;
  isNewMerchant: boolean;
  velocityLast10m: number;
  failedAttemptsLast24h: number;
  customerBaseline?: Partial<Customer>;
  merchantBaseline?: Partial<Merchant>;
  accountAgeDays?: number;
  timeOfDayHour?: number;
}

/**
 * AI Prototype Risk Engine (v1.4)
 * 
 * Computes dynamic transaction risk score [0..100], decomposes individual risk factor
 * contributions (interpretable feature weights), and generates human-readable reasoning.
 * Designed with pluggable interfaces for seamless transition to gradient-boosted or neural ML models.
 */
export function evaluateTransactionRisk(
  input: RiskEvaluationInput,
  weights: EngineWeights = DEFAULT_WEIGHTS,
  thresholds: EngineThresholds = DEFAULT_THRESHOLDS
): RiskScoreResult {
  const riskFactors: RiskFactor[] = [];
  let rawScore = 5; // Low baseline inherent network risk

  const avgAmount = input.customerBaseline?.averageAmount || 2500;
  const typicalRange: [number, number] = input.customerBaseline?.typicalAmountRange || [500, 5000];
  const usualLocation = input.customerBaseline?.usualLocations?.[0] || 'Hyderabad';
  const usualDevice = input.customerBaseline?.knownDevices?.[0] || 'Android Device';
  const typicalDailyTxn = input.customerBaseline?.typicalTxnPerDay || 3;

  // 1. Amount Deviation Factor
  const amountRatio = input.amount / Math.max(avgAmount, 100);
  if (amountRatio >= 20) {
    const contribution = Math.min(30, Math.round(weights.amountDeviationWeight * 1.3));
    rawScore += contribution;
    riskFactors.push({
      code: 'AMOUNT_MASSIVE_SPIKE',
      name: `Amount ${amountRatio.toFixed(0)}× customer average`,
      contribution,
      description: `Current payment (₹${input.amount.toLocaleString('en-IN')}) is ${amountRatio.toFixed(0)}x higher than customer typical average of ₹${avgAmount.toLocaleString('en-IN')}.`,
      category: 'AMOUNT',
      severity: 'CRITICAL',
    });
  } else if (amountRatio >= 5) {
    const contribution = Math.min(22, Math.round(weights.amountDeviationWeight * 0.9));
    rawScore += contribution;
    riskFactors.push({
      code: 'AMOUNT_HIGH_SPIKE',
      name: `Amount ${amountRatio.toFixed(1)}× customer average`,
      contribution,
      description: `Payment (₹${input.amount.toLocaleString('en-IN')}) significantly exceeds usual spending range (₹${typicalRange[0].toLocaleString('en-IN')}–₹${typicalRange[1].toLocaleString('en-IN')}).`,
      category: 'AMOUNT',
      severity: 'HIGH',
    });
  } else if (amountRatio >= 2.5) {
    const contribution = Math.round(weights.amountDeviationWeight * 0.45);
    rawScore += contribution;
    riskFactors.push({
      code: 'AMOUNT_MODERATE_SPIKE',
      name: 'Higher than usual amount',
      contribution,
      description: `Payment amount is moderately higher than usual baseline.`,
      category: 'AMOUNT',
      severity: 'MEDIUM',
    });
  }

  // 2. Device Novelty Factor
  if (input.isNewDevice) {
    const contribution = weights.newDeviceWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'DEVICE_UNRECOGNIZED',
      name: 'New / Unrecognized device fingerprint',
      contribution,
      description: 'Transaction originated from a previously unseen hardware/browser signature.',
      category: 'DEVICE',
      severity: 'HIGH',
    });
  }

  // 3. Location Anomaly / Impossible Travel
  const isLocationMismatch =
    Boolean(input.previousLocation && input.location !== input.previousLocation) ||
    (input.customerBaseline?.usualLocations &&
      !input.customerBaseline.usualLocations.includes(input.location));

  if (isLocationMismatch) {
    const contribution = weights.locationMismatchWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'LOCATION_ANOMALY',
      name: 'Unusual / Mismatched location',
      contribution,
      description: `Originating city (${input.location}) deviates from customer established activity center (${usualLocation}).`,
      category: 'LOCATION',
      severity: 'HIGH',
    });
  }

  // 4. Velocity Factor (Transactions in last 10 minutes)
  if (input.velocityLast10m >= 10) {
    const contribution = Math.min(30, Math.round(weights.velocityWeight * 1.4));
    rawScore += contribution;
    riskFactors.push({
      code: 'VELOCITY_EXTREME',
      name: 'Critical transaction velocity burst',
      contribution,
      description: `${input.velocityLast10m} payment attempts recorded within the last 10 minutes (Card testing or bot script signature).`,
      category: 'VELOCITY',
      severity: 'CRITICAL',
    });
  } else if (input.velocityLast10m >= 4) {
    const contribution = weights.velocityWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'VELOCITY_HIGH',
      name: 'High transaction velocity',
      contribution,
      description: `${input.velocityLast10m} transactions initiated in a short 10-minute window.`,
      category: 'VELOCITY',
      severity: 'HIGH',
    });
  } else if (input.velocityLast10m >= 2) {
    const contribution = Math.round(weights.velocityWeight * 0.4);
    rawScore += contribution;
    riskFactors.push({
      code: 'VELOCITY_MODERATE',
      name: 'Moderate transaction velocity',
      contribution,
      description: `Multiple transactions observed within recent minutes.`,
      category: 'VELOCITY',
      severity: 'MEDIUM',
    });
  }

  // 5. Failed Attempts in last 24h
  if (input.failedAttemptsLast24h >= 4) {
    const contribution = Math.min(25, Math.round(weights.failedAttemptsWeight * 1.3));
    rawScore += contribution;
    riskFactors.push({
      code: 'FAILED_ATTEMPTS_SPIKE',
      name: `${input.failedAttemptsLast24h} recent failed payment attempts`,
      contribution,
      description: `Repeated authorization declines or OTP timeouts prior to this successful transaction.`,
      category: 'BEHAVIOR',
      severity: 'HIGH',
    });
  } else if (input.failedAttemptsLast24h >= 2) {
    const contribution = Math.round(weights.failedAttemptsWeight * 0.6);
    rawScore += contribution;
    riskFactors.push({
      code: 'FAILED_ATTEMPTS_MODERATE',
      name: `${input.failedAttemptsLast24h} recent failed attempts`,
      contribution,
      description: `Prior transaction declines noted in the last 24 hours.`,
      category: 'BEHAVIOR',
      severity: 'MEDIUM',
    });
  }

  // 6. New Merchant Factor
  if (input.isNewMerchant) {
    const contribution = weights.newMerchantWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'MERCHANT_FIRST_TIME',
      name: 'First-time merchant interaction',
      contribution,
      description: `Customer has no prior payment history with this merchant.`,
      category: 'MERCHANT',
      severity: 'LOW',
    });
  }

  // 7. Merchant Category / Risk Rate
  if (input.merchantBaseline?.fraudRate && input.merchantBaseline.fraudRate > 3.0) {
    const contribution = weights.highRiskCategoryWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'MERCHANT_HIGH_RISK_PROFILE',
      name: 'High-risk merchant profile',
      contribution,
      description: `Merchant has an elevated fraud incidence rate (${input.merchantBaseline.fraudRate.toFixed(1)}%).`,
      category: 'MERCHANT',
      severity: 'HIGH',
    });
  }

  // 8. Account Age Factor
  const age = input.accountAgeDays ?? input.customerBaseline?.accountAgeDays ?? 365;
  if (age <= 7) {
    const contribution = 8;
    rawScore += contribution;
    riskFactors.push({
      code: 'ACCOUNT_NEWBORN',
      name: 'Newly created customer account (< 7 days)',
      contribution,
      description: 'Account was registered very recently, presenting limited historical baseline.',
      category: 'BEHAVIOR',
      severity: 'MEDIUM',
    });
  }

  // 9. Time of day anomaly (e.g. 1 AM to 4 AM)
  const hour = input.timeOfDayHour !== undefined ? input.timeOfDayHour : 14;
  if (hour >= 1 && hour <= 4 && (input.amount > 20000 || input.isNewDevice)) {
    const contribution = weights.timeAnomalyWeight;
    rawScore += contribution;
    riskFactors.push({
      code: 'TIME_ANOMALY',
      name: 'Odd-hour high-value transaction',
      contribution,
      description: `Transaction initiated during low-activity nighttime window (${hour}:00 hrs).`,
      category: 'BEHAVIOR',
      severity: 'LOW',
    });
  }

  // Clamp final score strictly between 0 and 100
  const finalScore = Math.min(100, Math.max(1, rawScore));

  // Determine Risk Level & Decision
  let riskLevel: RiskLevel = 'LOW';
  let decision: DecisionType = 'APPROVE';

  if (finalScore <= thresholds.approveMax) {
    riskLevel = 'LOW';
    decision = 'APPROVE';
  } else if (finalScore <= thresholds.verifyMax) {
    riskLevel = 'MEDIUM';
    decision = 'VERIFY';
  } else if (finalScore <= thresholds.holdMax) {
    riskLevel = 'HIGH';
    decision = 'HOLD';
  } else {
    riskLevel = 'CRITICAL';
    decision = 'BLOCK';
  }

  // Generate Explainable AI narrative
  const explanation = generateExplanation({
    riskScore: finalScore,
    decision,
    amountRatio,
    amount: input.amount,
    avgAmount,
    isNewDevice: input.isNewDevice,
    isLocationMismatch: Boolean(isLocationMismatch),
    location: input.location,
    usualLocation,
    velocityLast10m: input.velocityLast10m,
    failedAttempts: input.failedAttemptsLast24h,
    isNewMerchant: input.isNewMerchant,
    riskFactors,
  });

  return {
    riskScore: finalScore,
    riskLevel,
    decision,
    riskFactors: riskFactors.sort((a, b) => b.contribution - a.contribution),
    explanation,
    customerBaseline: {
      averageAmount: avgAmount,
      typicalRange,
      usualLocation,
      usualDevice,
      typicalDailyTxn,
    },
    metrics: {
      amountRatio,
      velocityLast10m: input.velocityLast10m,
      isNewDevice: input.isNewDevice,
      isLocationAnomaly: Boolean(isLocationMismatch),
      failedAttempts: input.failedAttemptsLast24h,
      isNewMerchant: input.isNewMerchant,
    },
  };
}

/**
 * Formulate clear, human-readable Explainable AI narrative.
 */
function generateExplanation(params: {
  riskScore: number;
  decision: DecisionType;
  amountRatio: number;
  amount: number;
  avgAmount: number;
  isNewDevice: boolean;
  isLocationMismatch: boolean;
  location: string;
  usualLocation: string;
  velocityLast10m: number;
  failedAttempts: number;
  isNewMerchant: boolean;
  riskFactors: RiskFactor[];
}): string {
  if (params.riskScore <= 30) {
    return `This transaction aligns closely with the customer's typical behavioral baseline. The transaction amount (₹${params.amount.toLocaleString('en-IN')}) is within expected limits, initiated from a recognized device in ${params.location} with normal velocity. Risk level is low and safe to auto-approve.`;
  }

  const reasons: string[] = [];

  if (params.amountRatio >= 5) {
    reasons.push(`the payment amount (₹${params.amount.toLocaleString('en-IN')}) is significantly higher (${params.amountRatio.toFixed(1)}×) than the customer's historical average of ₹${params.avgAmount.toLocaleString('en-IN')}`);
  } else if (params.amountRatio >= 2) {
    reasons.push(`the payment amount is moderately elevated compared to past transactions`);
  }

  if (params.isNewDevice) {
    reasons.push(`the transaction originated from a new, unrecognized device fingerprint`);
  }

  if (params.isLocationMismatch) {
    reasons.push(`the originating location (${params.location}) differs from the customer's usual activity hub (${params.usualLocation})`);
  }

  if (params.velocityLast10m >= 4) {
    reasons.push(`${params.velocityLast10m} rapid transactions occurred within the last 10 minutes`);
  }

  if (params.failedAttempts >= 2) {
    reasons.push(`there were ${params.failedAttempts} prior failed authorization attempts in the last 24 hours`);
  }

  if (params.isNewMerchant) {
    reasons.push(`this is the first transaction with this merchant`);
  }

  if (reasons.length === 0) {
    return `This transaction presents moderate statistical deviation across payment and merchant features, warranting a risk score of ${params.riskScore}/100 and a recommendation to ${params.decision}.`;
  }

  const joinedReasons = reasons.length === 1
    ? reasons[0]
    : reasons.slice(0, -1).join(', ') + ', and ' + reasons[reasons.length - 1];

  return `This transaction was flagged with a risk score of ${params.riskScore}/100 because ${joinedReasons}. The system recommends ${params.decision} pending risk analyst verification.`;
}

/**
 * Helper to evaluate simulation payload
 */
export function evaluateSimulation(
  sim: SimulationPayload,
  weights: EngineWeights = DEFAULT_WEIGHTS,
  thresholds: EngineThresholds = DEFAULT_THRESHOLDS
): RiskScoreResult {
  const hour = parseInt(sim.timeOfDay.split(':')[0] || '14', 10);

  return evaluateTransactionRisk(
    {
      amount: sim.amount,
      customerId: sim.customerId,
      merchantId: sim.merchantId,
      location: sim.location,
      previousLocation: sim.previousLocation,
      isNewDevice: sim.isNewDevice,
      isNewMerchant: sim.isNewMerchant,
      velocityLast10m: sim.velocityLast10m,
      failedAttemptsLast24h: sim.failedAttemptsLast24h,
      accountAgeDays: sim.accountAgeDays,
      timeOfDayHour: hour,
      customerBaseline: {
        id: sim.customerId,
        averageAmount: sim.averageCustomerTransaction,
        typicalAmountRange: [
          Math.round(sim.averageCustomerTransaction * 0.3),
          Math.round(sim.averageCustomerTransaction * 2.2),
        ],
        usualLocations: [sim.previousLocation || 'Hyderabad'],
        knownDevices: sim.isNewDevice ? ['Old Recognized Device'] : [sim.deviceType],
        typicalTxnPerDay: 3,
        accountAgeDays: sim.accountAgeDays,
      },
    },
    weights,
    thresholds
  );
}
